samples/drivers/stepper/generic/src/main.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 Jilay Sandeep Pandya.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/stepper/stepper_trinamic.h>
+
+/*
+ * Get button configuration from the devicetree sw0 alias. This is mandatory.
+ */
+#define SW0_NODE DT_ALIAS(sw0)
+#if !DT_NODE_HAS_STATUS_OKAY(SW0_NODE)
+#error "Unsupported board: sw0 devicetree alias is not defined"
+#endif
+static const struct gpio_dt_spec button = GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios, {0});
+static struct gpio_callback button_cb_data;
+
+const struct device *stepper = DEVICE_DT_GET(DT_ALIAS(stepper));
+
+enum stepper_mode {
+	STEPPER_MODE_PING_PONG_RELATIVE,
+	STEPPER_MODE_PING_PONG_ABSOLUTE,
+	STEPPER_MODE_ROTATE_CW,
+	STEPPER_MODE_ROTATE_CCW,
+};
+
+atomic_t stepper_mode = STEPPER_MODE_PING_PONG_RELATIVE;
+
+int32_t ping_pong_target_position =
+	CONFIG_STEPS_PER_REV * CONFIG_PING_PONG_N_REV * DT_PROP(DT_ALIAS(stepper), micro_step_res);
+
+K_SEM_DEFINE(stepper_generic_sem, 0, 1);
+
+void stepper_callback(const struct device *dev, const enum stepper_event event, void *user_data)
+{
+	switch (event) {
+	case STEPPER_EVENT_STEPS_COMPLETED:
+		k_sem_give(&stepper_generic_sem);
+		break;
+	default:
+		break;
+	}
+}
+
+void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	stepper_mode = atomic_get(&stepper_mode) + 1;
```

Comment:
Atomics need to updated atomically :) modifying it, then checking if the modified value is valid, then modifying it again is not atomic and thus requires lock. 

Or, atomic_get() the atomic to a temp var, check it, then atomic_set() or atomic_inc()